src/VisualStudio/Core/Impl/Options/FullSolutionAnalysisOptionBinding.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Options;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Options
+{
+    internal class FullSolutionAnalysisOptionBinding
+    {
+        private readonly IOptionService _optionService;
+        private readonly string _languageName;
+
+        private readonly Option<bool> _fullSolutionAnalysis;
+        private readonly PerLanguageOption<bool> _closedFileDiagnostics;
+
+        public FullSolutionAnalysisOptionBinding(IOptionService optionService, string languageName)
+        {
+            _optionService = optionService;
+            _languageName = languageName;
+
+            _fullSolutionAnalysis = RuntimeOptions.FullSolutionAnalysis;
+            _closedFileDiagnostics = ServiceFeatureOnOffOptions.ClosedFileDiagnostic;
+        }
+
+        public bool Value
+        {
+            get
+            {
+                return _optionService.GetOption(_closedFileDiagnostics, _languageName) &&
+                       _optionService.GetOption(_fullSolutionAnalysis);
+            }
+
+            set
+            {
+                var oldOptions = _optionService.GetOptions();
+
+                // set normal option first
+                var newOptions = oldOptions.WithChangedOption(_closedFileDiagnostics, _languageName, value);
+
+                // we only enable this option if it is disabled. we never disable this option here.
```

Comment:
the runtime option will only be disabled by us. if user explicit disable the option, it will disable user explicit option (CloseFileDiagnostics) and for user end result is same.

we do this so that we have 2 options one controlled by user and persisted, one controlled by us and not persisted.

if user has full analysis on, and we disabled it, user can go and explicit enable it again or if user restart VS, full analysis will be back on.

basically we show one aggregated option to users, and internally it is combination of option user controlling and option we controlling.